// @sitecore-jss/sitecore-jss-dev-tools/dist/cjs/templating/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMetadata = exports.strip = exports.scaffoldFile = exports.ModuleType = exports.generatePlugins = exports.getComponentList = void 0;
var components_1 = require("./components");
Object.defineProperty(exports, "getComponentList", { enumerable: true, get: function () { return components_1.getComponentList; } });
var plugins_1 = require("./plugins");
Object.defineProperty(exports, "generatePlugins", { enumerable: true, get: function () { return plugins_1.generatePlugins; } });
Object.defineProperty(exports, "ModuleType", { enumerable: true, get: function () { return plugins_1.ModuleType; } });
var scaffold_1 = require("./scaffold");
Object.defineProperty(exports, "scaffoldFile", { enumerable: true, get: function () { return scaffold_1.scaffoldFile; } });
var strip_1 = require("./strip");
Object.defineProperty(exports, "strip", { enumerable: true, get: function () { return strip_1.strip; } });
var metadata_1 = require("./metadata");
Object.defineProperty(exports, "getMetadata", { enumerable: true, get: function () { return metadata_1.getMetadata; } });
