// @sitecore-jss/sitecore-jss-cli/dist/cjs/cli.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPackageScriptCommands = getPackageScriptCommands;
exports.makeCommand = makeCommand;
exports.default = default_1;
const cli_shared_1 = __importDefault(require("./cli-shared"));
const resolve_package_1 = __importDefault(require("./resolve-package"));
const run_package_script_1 = __importDefault(require("./run-package-script"));
const commands = __importStar(require("./scripts"));
/**
 * Get package script commands
 */
function getPackageScriptCommands() {
    return __awaiter(this, void 0, void 0, function* () {
        const packageJson = yield (0, resolve_package_1.default)();
        const result = {};
        if (!packageJson || !packageJson.scripts) {
            return result;
        }
        Object.keys(packageJson.scripts).forEach((script) => {
            if (script === 'jss') {
                return;
            }
            const command = makeCommand(script);
            result[script] = command;
        });
        return result;
    });
}
/**
 * @param {string} script script name
 */
function makeCommand(script) {
    return {
        command: script,
        describe: 'package.json script',
        builder: {},
        disableStrictArgs: true,
        handler: (argv) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            if (argv._[0]) {
                (0, run_package_script_1.default)(process.argv.slice(2));
            }
        },
    };
}
/**
 * implements CLI commands when executed from a local node_modules folder
 */
function default_1() {
    return __awaiter(this, void 0, void 0, function* () {
        const packageScripts = yield getPackageScriptCommands();
        (0, cli_shared_1.default)(Object.assign({}, commands, packageScripts));
    });
}
